### PR TITLE
IOTEDGE-691 - Create the Client App Dev Guide structure

### PR DESCRIPTION
**Pull request by lanafrost**
This commit sets up the Client App Dev Guide and walks users through the first two examples of static and dynamic initialisation.

### CODE REVIEW

File: docs/client-app-dev-guide.md [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+### Before You Start
+Before you start developing clients, it is helpful to have an overview of the IEC and its 
+components. Read the official [IEC Documentation](http://backstage.forgerock.com/docs/iec/6.5/) 
+as a starting point.
+
+The code examples in this guide assume that you are using the 
+[IEC training environment](https://github.com/ForgeRock/identity-edge-controller-docs/tree/master/training). 
+If you are not using the training environment, make sure that you have installed all the 
+required components, including the IEC SDK, and that everything is up and running, as 
+described in the [IEC Installation Guide](http://backstage.forgerock.com/docs/iec/6.5/install-guide).
```
Comment (jgadsden) [inline docs/client-app-dev-guide.md]: The link http://backstage.forgerock.com/docs/iec/6.5/install-guide does not work for me?

Comment (lanafrost) [reply docs/client-app-dev-guide.md]: Yes, that's what we're waiting for Zoli's team to do. I could actually publish these docs to Backstage - the direct links will then work, although nobody will be able to find the docs without the navigation.

**Comment by jgadsden** [comment]
I guess once the iec area of http://backstage.forgerock.com/docs/ is created then the links such as:
http://backstage.forgerock.com/docs/iec
will work OK

File: docs/client-app-dev-guide.md [hunk 2]
```diff
@@ -0,0 +1,233 @@
+## ForgeRock Identity Edge Application Developers Guide
+
+This guide shows you how to use the ForgeRock® Identity Edge Controller (IEC) SDK to 
+develop client applications and to register them with the IEC Service. The IEC SDK client 
+library provides a simple C API for client applications to invoke ForgeRock Access Manager
```
Comment (LimaEchoCharlie) [inline docs/client-app-dev-guide.md]: We have a C and Go client API. The examples are in C but we should probably mention Go here as well.

File: docs/client-app-dev-guide.md [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+types will fail if the IEC Service has not registered successfully with AM. In general, the 
+functional flow is from the SDK Client library (`libiecclient.so`) API, via ZMQ and the IEC 
+Service to the AM IEC Plugin and back.
+
+For details of the IEC components and how they interact, see the 
+[IEC Getting Started Guide](http://backstage.forgerock.com/docs/iec/6.5/getting-started).
+
+Log in to the AM Admin Console. The default credentials for the AM admin user in the training 
+environment are `amadmin` and `password`. Your AM instance must be configured with the 
+`edge` realm.
```
Comment (LimaEchoCharlie) [inline docs/client-app-dev-guide.md]: Should we say that the training environment comes with the `edge` realm pre-configured?

File: docs/client-app-dev-guide.md [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+         "compress_file_backups": true
+       },
+       "client_configuration": {
+         "id": "iec-client"
+       }
+     }
+  ```
+  The configuration database *must* exist for the ZMQ communication to be established. Once 
+  the database exists, you can update it from AM through the IEC Service. For security 
+  reasons, you should delete or protect the database once the client has been initialized.
```
Comment (LimaEchoCharlie) [inline docs/client-app-dev-guide.md]: database -> JSON configuration file

File: docs/client-app-dev-guide.md [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+     }
+  ```
+  The configuration database *must* exist for the ZMQ communication to be established. Once 
+  the database exists, you can update it from AM through the IEC Service. For security 
+  reasons, you should delete or protect the database once the client has been initialized. 
+  
+* Dynamically, using the `iec_set_attribute` function to set individual attribute values.
+
+  Dynamic configuration takes precedence over a database configuration, if one exists. You 
+  can use the `iec_set_attribute` utility to set the following attribute values:
```
Comment (LimaEchoCharlie) [inline docs/client-app-dev-guide.md]: We can refer the dev to `libiectypes.h` for the full list.

File: docs/client-app-dev-guide.md [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+  ZMQMessageTimeoutSec
+  LoggingEnabled
+  LoggingDebug
+  LoggingLogfile
+  LoggingMaxSize
+  LoggingMaxBackups
+  LoggingMaxAge
+  LoggingCompressBackups
+  ClientID
+  ```
```
Comment (LimaEchoCharlie) [inline docs/client-app-dev-guide.md]:
```
    IEC_ENDPOINT,                   // string
    IEC_SECRETKEY,                  // string
    IEC_PUBLICKEY,                  // string
    IEC_SERVERPUBLICKEY,            // string
    IEC_MSGTIMEOUTSEC,              // int
    IEC_CLIENT_ID,                  // string
    IEC_LOGGING_ENABLED,            // bool
    IEC_LOGGING_DEBUG,              // bool
    IEC_LOGGING_LOGFILE,            // string
    IEC_LOGGING_MAXSIZE,            // int
    IEC_LOGGING_MAXBACKUPS,         // int
    IEC_LOGGING_MAXAGE,             // int
    IEC_LOGGING_COMPRESSBACKUPS,    // bool
```

File: docs/client-app-dev-guide.md [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+   ~/forgerock/iecutil -file sdk-config.json -initialise sdk
+   iec util: Initialising sdk
+   iec util: Finished sdk initialisation
+   ```
+   
+   An SDK application looks in the directory from which it's run for a configuration database or for a 
+   file that contains the database location.
+   
+   **Note** If you change the configuration and need to reinitialize the SDK, remove the 
+  `iec-sdk.db` in the application directory, then run the initialization again.
```
Comment (LimaEchoCharlie) [inline docs/client-app-dev-guide.md]: 👍

File: docs/client-app-dev-guide.md [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+   ./init_sdk_static
+   ```
+   
+   If the initialization was successful, you should see the `iec-client` identity in the AM Console.
+
+#### Example 2: Initializing a Client Dynamically
+
+If you are running this example after the previous example:
+
+* Delete the `iec-client` identity in the AM Console.
```
Comment (LimaEchoCharlie) [inline docs/client-app-dev-guide.md]: As the dynamic client has a different id `iec-dynamic-client` to the static client, it isn't required to delete the the `iec-dynamic-client` identity.

File: docs/client-app-dev-guide.md (hunk 6 above)
Comment (lanafrost) [reply docs/client-app-dev-guide.md]: I removed the list, provided two examples, and a link to `libiectypes.h`.

**Pull request merged**